compiler/luci/logex/src/CircleNodeSummaryBuilder.h

```diff
@@ -0,0 +1,104 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __LUCI_LOGEX_CIRCLE_NODE_SUMMARY_BUILDER__
+#define __LUCI_LOGEX_CIRCLE_NODE_SUMMARY_BUILDER__
+
+#include "CircleNodeSummaryBuilderHelper.h"
+
+#include <luci/IR/CircleDialect.h>
+#include <luci/IR/CircleNode.h>
+#include <locop/NodeSummary.h>
+#include <locop/SymbolTable.h>
+
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace luci
+{
+
+class CircleNodeSummaryBuilder
+{
+public:
+  bool build(const loco::Node *node, const locop::SymbolTable *tbl, locop::NodeSummary &s)
+  {
+    if (node->dialect() != luci::CircleDialect::get())
+      return false;
+
+    auto ptr_to_str = [](const void *ptr) {
+      std::stringstream ss;
+      ss << ptr;
+      return ss.str();
+    };
+
+    auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+    if (const auto builder = create_builder(circle_node))
+    {
+      if (!builder->validate(circle_node))
+      {
+        s.state(locop::NodeDesc::State::Invalid);
+        return false;
+      }
+
+      auto input_names = builder->get_input_names(circle_node);
+      assert(node->arity() == input_names.size());
+      for (uint32_t i = 0; i < node->arity(); ++i)
+        s.args().append(input_names.at(i), tbl->lookup(node->arg(i)));
+
+      builder->build_attributes(circle_node, tbl, s);
+      builder->update_status(s);
+
+      s.opname(circle_opname(circle_node->opcode()));
+      s.comments().append("[" + circle_node->name() + "] = " + ptr_to_str(node));
+
+      return true;
+    }
+    else
+    {
+      // When SummaryBuilder is not implemented, return false
+      return false;
+    }
+  }
+
+  /**
+   * @brief Template methods for building node summary.
+   *        Default behavior is building a node with no inputs.
+   */
+private:
+  virtual bool validate(const luci::CircleNode *node) { return true; }
+
+  virtual std::vector<std::string> get_input_names(const luci::CircleNode *node)
+  {
+    // Return empty names for default
+    return std::vector<std::string>();
+  }
+
+  virtual void build_attributes(const luci::CircleNode *node, const locop::SymbolTable *tbl,
+                                locop::NodeSummary &s)
+  {
+    // Do nothing for default
+  }
+
+  virtual void update_status(locop::NodeSummary &s) { s.state(locop::NodeDesc::State::Complete); }
+
+private:
+  std::shared_ptr<CircleNodeSummaryBuilder> create_builder(const luci::CircleNode *node);
```

Comment:
Can we use `unique_ptr` here? (I'm not good at using `unique_ptr` vs `shared_ptr`)

Reply:
Yeap, we can use `unique_ptr` :)
I will replace with it